Best Practices/Clean Code.md

```diff
@@ -91,75 +82,880 @@ In other words, the Law of Demeter principle states that a module (class) should
+
+    public CustomerService(ICustomerRepository customerRepository)
+    {
+        _customerRepository = customerRepository;
+    }
+
+    public void AddCustomer(Customer customer)
+    {
+        _customerRepository.Add(customer);
+    }
+}
+
+public interface ICustomerRepository
+{
+    void Add(Customer customer);
+}
+
+public class SqlCustomerRepository : ICustomerRepository
+{
+    public void Add(Customer customer)
+    {
+        // SQL-specific logic for adding a customer
+    }
+}
+```
+
+The `CustomerService` class depends on an abstraction (`ICustomerRepository`) rather than a concrete class. The dependency is injected into the `CustomerService` constructor, promoting loose coupling and adherence to the DIP. The benefits of this approach are the following:
+* Decoupling the high-level `CustomerService` from the low-level `SqlCustomerRepository` so it can work with any implementation of `ICustomerRepository`.
+* Simplification of data storage switching (e.g., from SQL to a NoSQL database) without modifying the `CustomerService` class.
+* Easier writing of unit tests for `CustomerService` because the `ICustomerRepository` interface can be mocked, allowing `CustomerService` to be tested in isolation from the data layer.
+
+### Don't Repeat Yourself (DRY)
+
+The DRY principle promotes reusable component creation so that the code responsible for one thing appears only in one place. It suggests avoiding repetition by replacing duplicate logic or code snippets with shared methods or classes. That way, code is easier to maintain and update, and there is a smaller chance of errors and unwanted program behavior.
+
+**Good example:** 
+
+```c#
+public interface IShape
+{
+    public double CalculateArea();
+}
+
+public class Rectangle : IShape
+{
+    public double CalculateArea()
+    {
+        // Calculation logic for rectangle's area
+    }
+}
+
+public class Circle : IShape
+{
+    public double CalculateArea()
+    {
+        // Calculation logic for circle's area
+    }
+}
+```
+
+We've created a method `Area` in an interface `IShape`, so there is no code duplication for calculating areas in each subclass (`Rectangle` and `Circle`). Instead, each subclass only needs to implement the logic specific to its shape.
+
+**Bad example** (repeating code for area calculation)**:**
+
+```c#
+public interface IShape
+{
+    public double CalculateArea();
+}
+
+public class Rectangle : IShape
+{
+    public double CalculateArea()
+    {
+        // Calculation logic for rectangle's area
+    }
+}
+
+public class Circle : IShape
+{
+    public double CalculateArea()
+    {
+        // Calculation logic for circle's area
+    }
+}
+
+public class AreaCalculator
+{
+    public double CalculateRectangleArea(double length, double width)
+    {
+        return length * width;
+    }
+
+    public double CalculateCircleArea(double radius)
+    {
+        return Math.PI * radius * radius;
+    }
+}
+```
```

Comment:
Even though this example is not wrong, I think we could think of a more realistic example we encounter in our everyday work.

One of the most common mistakes people make is when a specific action requires a few steps to perform. Common examples are role validations that throw an exception, sending notifications that require initializing a few DTO classes, so I suggest we show something like that here :)